File: src/game/state.py
"""
游戏状态：双方原子池、生命、场地（各 3 格）、当前玩家与回合阶段等。
"""
from typing import Dict, List, Optional, Set, Tuple
from src.grid.cell import Cell, ATOM_BLACK, ATOM_RED, ATOM_BLUE, ATOM_GREEN, COLORS as ATOM_COLORS

GridPoint = Tuple[int, int]
from src.game.game_config import GameConfig, default_config

# 阶段
PHASE_CONFIRM = 0   # 确认场上状态，选 a/b/c
PHASE_DRAW = 1      # 获取原子
PHASE_PLACE = 2     # 排布
PHASE_ACTION = 3    # 动作（进攻与效果）

# 阶段 0 选项
CHOICE_EXTRA_DRAW = "a"   # 本回合获得 5+x 原子
CHOICE_EXTRA_PLACE = "b"  # 本回合最多放置 4+x 原子
CHOICE_EXTRA_ATTACK = "c" # 本回合可进攻 x 次

INITIAL_HP = 20
INITIAL_POOL = {ATOM_BLACK: 7, ATOM_RED: 1, ATOM_BLUE: 1, ATOM_GREEN: 1}


def make_initial_pool() -> Dict[str, int]:
    return dict(INITIAL_POOL)


def make_cells() -> List[Cell]:
    from src.config import (
        DEFAULT_GRID_ROWS,
        DEFAULT_GRID_COLS,
        GRID_CENTER_R,
        GRID_CENTER_C,
        HEX_RADIUS,
    )
    return [
        Cell(
            DEFAULT_GRID_ROWS,
            DEFAULT_GRID_COLS,
            center_r=GRID_CENTER_R,
            center_c=GRID_CENTER_C,
            hex_radius=HEX_RADIUS,
        )
        for _ in range(3)
    ]


class GameState:
    def __init__(self, config: Optional[GameConfig] = None):
        cfg = config or default_config()
        self.config = cfg
        # 玩家 0 与 1：原子池（按配置）、生命、3 个格子
        pool_copy = dict(cfg.initial_pool)
        self.pools = [dict(pool_copy), dict(pool_copy)]
        self.hp: List[int] = [INITIAL_HP, INITIAL_HP]
        self.cells: List[List[Cell]] = [make_cells(), make_cells()]

        self.current_player: int = 0
        self.phase: int = PHASE_CONFIRM
        self.phase_0_choice: Optional[str] = None  # "a" | "b" | "c"

        self.base_draw_count: int = cfg.base_draw_count
        self.base_place_limit: int = cfg.base_place_limit
        self.draw_weights: List[int] = list(cfg.draw_weights)

        # 本回合限制（在阶段 0 选择后设定）
        self.turn_draw_count: int = cfg.base_draw_count
        self.turn_place_limit: int = cfg.base_place_limit
        self.turn_attack_limit: int = 1

        # 本回合已用
        self.turn_placed_count: int = 0
        self.turn_attack_used: int = 0

        # 先手第一回合不能进攻
        self.turn_number: int = 0
        self.is_first_turn: bool = True

        # 蓝效果：与该蓝相邻的黑原子在下一回合内不可被破坏。(player -> set of (cell_i, (r,c)))
        self.blue_protected_points: Dict[int, Set[Tuple[int, GridPoint]]] = {0: set(), 1: set()}
        self.blue_protection_until_turn: Dict[int, int] = {}  # player -> 保护持续到该回合号（不含）

    def is_black_protected(self, player: int, cell_i: int, pt: GridPoint) -> bool:
        """该玩家的该格该格点上的黑原子是否处于蓝效果保护中。"""
        s = self.blue_protected_points.get(player)
        if not s:
            return False
        return (cell_i, pt) in s

    def opponent(self, player: int) -> int:
        return 1 - player

    def pool(self, player: int) -> Dict[str, int]:
        return self.pools[player]

    def player_cells(self, player: int) -> List[Cell]:
        return self.cells[player]

    def non_empty_cell_count(self, player: int) -> int:
        return sum(1 for c in self.player_cells(player) if not c.is_empty())

    def cells_with_black_count(self, player: int) -> int:
        """有至少一个黑原子的格子数量；每回合该玩家可进攻次数即为此数。"""
        return sum(1 for c in self.player_cells(player) if c.has_black())

    def x_for_turn(self) -> int:
        """当前回合玩家的 x（非空格子数）。"""
        return self.non_empty_cell_count(self.current_player)

    def can_attack_this_turn(self) -> bool:
        if self.turn_attack_used >= self.turn_attack_limit:
            return False
        if self.is_first_turn:
            return False
        return True

    def winner(self) -> Optional[int]:
        """0 或 1 表示胜者，None 表示未结束。"""
        if self.hp[0] <= 0:
            return 1
        if self.hp[1] <= 0:
            return 0
        return None

File: web/src/game/state.js
/**
 * Game state: pools, HP, cells, phase, etc.
 */
import { Cell } from './cell.js'
import {
  ATOM_BLACK,
  ATOM_RED,
  ATOM_BLUE,
  ATOM_GREEN,
  ATOM_YELLOW,
  INITIAL_HP,
  INITIAL_POOL,
  PHASE_CONFIRM,
  PHASE_DRAW,
  PHASE_PLACE,
  PHASE_ACTION,
} from './config.js'

const GRID_ROWS = 100
const GRID_COLS = 100
const CENTER_R = 50
const CENTER_C = 50
const HEX_RADIUS = 15

function makeCells(count) {
  const n = Math.max(2, Math.min(6, count ?? 3))
  return Array.from({ length: n }, () => new Cell(GRID_ROWS, GRID_COLS, CENTER_R, CENTER_C, HEX_RADIUS))
}

export function createGameState(config = {}) {
  const initialPool = { ...INITIAL_POOL }
  const cfg = {
    baseDrawCount: 10,
    basePlaceLimit: 10,
    drawWeights: [3, 1, 1, 1, 1, 0, 0],
    initialHp: INITIAL_HP,
    cellCount: 3,
    ...config,
  }
  const base = [3, 1, 1, 1, 1, 0, 0, 0]
  const numColors = 8
  const wIn = Array.isArray(cfg.drawWeights) ? cfg.drawWeights : base
  const weights = wIn.length >= numColors ? wIn.slice(0, numColors) : [...wIn, ...Array(numColors - wIn.length).fill(0)].slice(0, numColors)
  const hpVal = Math.max(1, Math.min(99, cfg.initialHp ?? INITIAL_HP))
  const cellCount = Math.max(2, Math.min(6, cfg.cellCount ?? 3))
  return {
    config: cfg,
    pools: [{ ...initialPool }, { ...initialPool }],
    hp: [hpVal, hpVal],
    cells: [makeCells(cellCount), makeCells(cellCount)],
    currentPlayer: 0,
    phase: PHASE_CONFIRM,
    phase0Choice: null,
    baseDrawCount: cfg.baseDrawCount,
    basePlaceLimit: cfg.basePlaceLimit,
    drawWeights: weights,
    turnDrawCount: cfg.baseDrawCount,
    turnPlaceLimit: cfg.basePlaceLimit,
    turnAttackLimit: 1,
    turnPlacedCount: 0,
    turnAttackUsed: 0,
    turnNumber: 0,
    attackedCellsThisTurn: [],
    attackedEnemyCellIndicesThisTurn: [],
    redEffectTargetCellIndicesThisTurn: [],
    isFirstTurn: true,
    blueProtectedPoints: { 0: new Set(), 1: new Set() },
    blueProtectionUntilTurn: {},
    yellowPriorityPoints: { 0: new Set(), 1: new Set() },
    yellowPriorityUntilTurn: {},
    graySilencedPoints: { 0: new Set(), 1: new Set() },
    graySilencedUntilTurn: {},
    placementHistory: [],
  }
}

export function opponent(state, player) {
  return 1 - player
}

export function pool(state, player) {
  return state.pools[player]
}

export function playerCells(state, player) {
  return state.cells[player]
}

export function cellsWithBlackCount(state, player) {
  return state.cells[player].filter((c) => c.hasBlack()).length
}

export function nonEmptyCellCount(state, player) {
  return state.cells[player].filter((c) => !c.isEmpty()).length
}

export function xForTurn(state) {
  return nonEmptyCellCount(state, state.currentPlayer)
}

export function canAttackThisTurn(state) {
  if (state.turnAttackUsed >= state.turnAttackLimit) return false
  if (state.isFirstTurn) return false
  return true
}

/** 该格本回合是否已进攻过（每格每回合只能进攻一次） */
export function hasCellAttackedThisTurn(state, player, cellIndex) {
  const list = state.attackedCellsThisTurn ?? []
  return list.some(([p, ci]) => p === player && ci === cellIndex)
}

export function winner(state) {
  if (state.hp[0] <= 0) return 1
  if (state.hp[1] <= 0) return 0
  return null
}

export function isBlackProtected(state, player, cellIndex, pt) {
  const pk = typeof pt === 'string' ? pt : `${pt[0]},${pt[1]}`
  return state.blueProtectedPoints[player]?.has(`${cellIndex}:${pk}`) ?? false
}

export function isBlackYellowPriority(state, player, cellIndex, pt) {
  const pk = typeof pt === 'string' ? pt : `${pt[0]},${pt[1]}`
  return state.yellowPriorityPoints[player]?.has(`${cellIndex}:${pk}`) ?? false
}

/** 该格点是否处于灰原子点击效果的沉默区域内（下一回合内不可发动其他原子点击效果） */
export function isGraySilenced(state, player, cellIndex, pt) {
  const pk = typeof pt === 'string' ? pt : `${pt[0]},${pt[1]}`
  const until = state.graySilencedUntilTurn?.[player]
  if (until == null || state.turnNumber >= until) return false
  return state.graySilencedPoints[player]?.has(`${cellIndex}:${pk}`) ?? false
}

/** 返回该玩家有黄原子的格子索引列表 */
export function cellsWithYellow(state, player) {
  const out = []
  for (let i = 0; i < state.cells[player].length; i++) {
    if (state.cells[player][i].hasYellow()) out.push(i)
  }
  return out
}

/** 某格黄原子个数（用于黄持续效果：进攻顺序） */
export function yellowCountInCell(cell) {
  return cell.countByColor?.()?.[ATOM_YELLOW] ?? 0
}

/**
 * 进攻时可选的对方格子索引。
 * 黄原子持续效果：对方若要进攻我方某格（该格有 x 个黄原子），须已进攻过所有「黄原子数严格大于 x」的我方格子。
 * 因此按黄原子数从多到少依次可被选为攻击目标。
 */
/** 有灰原子的格子内黄持续效果无效 */
function effectiveYellowCountInCell(cell) {
  if (cell.hasGray?.()) return 0
  return yellowCountInCell(cell)
}

export function getAttackableEnemyCellIndices(state) {
  const opp = 1 - state.currentPlayer
  const cells = state.cells[opp]
  const n = cells.length
  const attackedSet = new Set(state.attackedEnemyCellIndicesThisTurn ?? [])
  const yellowCounts = cells.map((c) => effectiveYellowCountInCell(c))

  const candidateIndices = Array.from({ length: n }, (_, i) => i).filter(
    (i) => !cells[i].isEmpty() && cells[i].hasBlack()
  )
  if (candidateIndices.length === 0) return []

  const attackable = candidateIndices.filter((i) => {
    const x = yellowCounts[i]
    for (let j = 0; j < n; j++) {
      if (yellowCounts[j] > x && !attackedSet.has(j)) return false
    }
    return true
  })
  return attackable
}

/**
 * 红效果可选的对方格子索引。
 * 与进攻相同顺序约束：对方某格有 x 个黄原子时，须已对「黄原子数严格大于 x」的我方其他格子发动过红效果后才能选该格。
 */
export function getRedEffectTargetableEnemyCellIndices(state) {
  const cur = state.currentPlayer
  const opp = 1 - cur
  const cells = state.cells[opp]
  const n = cells.length
  const doneSet = new Set(state.redEffectTargetCellIndicesThisTurn ?? [])
  const yellowCounts = cells.map((c) => effectiveYellowCountInCell(c))

  const candidateIndices = Array.from({ length: n }, (_, i) => i).filter(
    (i) => !cells[i].isEmpty() && cells[i].hasBlack()
  )
  if (candidateIndices.length === 0) return []

  return candidateIndices.filter((i) => {
    const x = yellowCounts[i]
    for (let j = 0; j < n; j++) {
      if (yellowCounts[j] > x && !doneSet.has(j)) return false
    }
    return true
  })
}
